internal/vm/ioproxy.go

```diff
@@ -0,0 +1,214 @@
+//
+// The provided context is canceled if the process fails to create/exec or
+// when the process exits after a successful create/exec.
+type IOConnector func(procCtx context.Context, logger *logrus.Entry) <-chan IOConnectorResult
+
+// IOConnectorResult represents the result of attempting to establish an IO
+// connection. If successful, the ReadWriteCloser should be non-nil and Err
+// should be nil. If unsuccessful, ReadWriteCloser should be nil and Err
+// should be non-nil.
+type IOConnectorResult struct {
+	io.ReadWriteCloser
+	Err error
+}
+
+// IOConnectorPair holds the read and write side of IOConnectors whose IO
+// should be proxied.
+type IOConnectorPair struct {
+	ReadConnector  IOConnector
+	WriteConnector IOConnector
+}
+
+func (connectorPair *IOConnectorPair) proxy(
+	proc *vmProc,
+	logger *logrus.Entry,
+	timeoutAfterExit time.Duration,
+) (ioInitDone <-chan error, ioCopyDone <-chan error) {
+	initDone := make(chan error, 2)
+	copyDone := make(chan error)
+
+	// Start the initialization process. Any synchronous setup made by the connectors will
+	// be completed after these lines. Async setup will be done once initDone is closed in
+	// the goroutine below.
+	readerResultCh := connectorPair.ReadConnector(proc.ctx, logger.WithField("direction", "read"))
+	writerResultCh := connectorPair.WriteConnector(proc.ctx, logger.WithField("direction", "write"))
+
+	go func() {
+		defer close(copyDone)
+
+		var reader io.ReadCloser
+		var writer io.WriteCloser
+		var ioInitErr error
+
+		// Send the first error we get to initDone, but consume both so we can ensure both
+		// end up closed in the case of an error
+		for readerResultCh != nil || writerResultCh != nil {
+			select {
+			case readerResult := <-readerResultCh:
+				readerResultCh = nil
+				reader, ioInitErr = readerResult.ReadWriteCloser, readerResult.Err
+			case writerResult := <-writerResultCh:
+				writerResultCh = nil
+				writer, ioInitErr = writerResult.ReadWriteCloser, writerResult.Err
+			}
+
+			if ioInitErr != nil {
+				logger.WithError(ioInitErr).Error("error initializing io")
+				initDone <- ioInitErr
+			}
+		}
+
+		close(initDone)
+		if ioInitErr != nil {
+			logClose(logger, reader, writer)
+			return
+		}
+
+		// IO streams have been initialized successfully
+
+		// Once the proc exits, wait the provided time before forcibly closing io streams.
+		// If the io streams close on their own before the timeout, the Close calls here
+		// should just be no-ops.
+		go func() {
+			<-proc.ctx.Done()
+			time.AfterFunc(timeoutAfterExit, func() {
+				logClose(logger, reader, writer)
+			})
+		}()
+
+		logger.Debug("begin copying io")
+		defer logger.Debug("end copying io")
+
+		_, err := io.CopyBuffer(writer, reader, make([]byte, internal.DefaultBufferSize))
+		if err != nil {
+			logger.WithError(err).Error("error copying io")
+			copyDone <- err
+		}
+	}()
+
+	return initDone, copyDone
+}
+
+type ioConnectorSet struct {
+	stdin  *IOConnectorPair
```

Comment:
Can this be an array of `IOConnectorPair`s instead?

Reply:
In theory sure but you would need to specify still that index 0 of the slice is stdin, index 1 is stdout, index 2 is stderr because there can be branching behavior for the different streams, i.e: https://github.com/sipsma/firecracker-containerd/blob/611800875f31dbb4f210561017a0b3e4ff2e19f5/internal/vm/ioproxy.go#L164

Given that, it seems like a lot more straightforward to just have 3 specific fields. I think it's highly unlikely we would want to support more logging FDs in the future in this context, so our code shouldn't need to handle arbitrary numbers.

Reply:
That branching logic could still work with an array, no? All you care about is the proxy function and passing in a name, which could be stored on the connector set.

Reply:
Sure, but what does that get us relative to just using a field for each of std<in/out/err>? It seems like it is just a more roundabout way of specifying them

Reply:
It cleans up the code, so you can put it in a for loop. It generalizes it making it easier to add things in the future, if needed.

Reply:
> It generalizes it making it easier to add things in the future, if needed.

At this point, I don't really anticipate needing additional IO streams or FDs passed to exec'd processes.  Arbitrary FDs would be fairly difficult across a vsock.

Named fields here are clearer in the code that uses them; you can see immediately that it's `stderr` instead of `io[2]` and need to remember the mapping of file descriptors to their common name.

Reply:
I can make a for loop now with
```
for stream := range []*IOConnectorPair{set.stdin, set.stdout, set.stderr} {
   ...
}
```

but then I need to add more fields to the struct (for the name of stream and what timeouts to use) just to save a total of probably ~10 lines of code. I agree with Sam, I don't feel like that results in more clarity.

Reply:
It's just for setup and the code is identical between the different io streams. I'm more for putting this in an array, as I think having them named doesn't yield anything in this context. However, since there seems to be more in favor of keeping it the way it is, I'll go ahead and approve this PR.